### PR TITLE
feat: add oidc option for sso

### PR DESCRIPTION
**Pull request by corang**
## Description

...

## Related Issue

Fixes #

Relates to #

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Other (security config, docs update, etc)

## Checklist before merging

- [ ] Test, docs, adr added or updated as needed
- [ ] [Contributor Guide Steps](https://github.com/defenseunicorns/uds-package-mattermost/blob/main/CONTRIBUTING.md#developer-workflow) followed

### CODE REVIEW

File: chart/values.yaml [hunk 1]
```diff
@@ -29,13 +29,19 @@ postgres:
...
   secretName: mattermost-sso
 
   # These should typically be disabled if SSO is enabled
   enable_sign_up_with_email: false
   enable_sign_in_with_email: false
   enable_sign_in_with_username: false
 
+  # Config for OIDC/SAML
+  login_button_text: "Login with SSO"
+  login_button_color: "#145DBF"
```
Comment (Racer159) [inline chart/values.yaml:43]: This should default to the Defense Unicorns blue as defined in our brand guide.

```suggestion
  login_button_color: "#144A8F"
```

File: chart/templates/uds-package.yaml [hunk 2]
```diff
@@ -14,16 +14,28 @@ spec:
...
+        MM_OPENIDSETTINGS_ENABLE: "false"
         MM_GITLABSETTINGS_ENABLE: "{{ .Values.sso.enabled | toString }}"
         MM_GITLABSETTINGS_ID: "clientField(clientId)"
         MM_GITLABSETTINGS_SECRET: "clientField(secret)"
         MM_GITLABSETTINGS_AUTHENDPOINT: "https://sso.{{ .Values.domain }}/realms/uds/protocol/openid-connect/auth"
         MM_GITLABSETTINGS_TOKENENDPOINT: "http://keycloak-http.keycloak.svc.cluster.local:8080/realms/uds/protocol/openid-connect/token"
         MM_GITLABSETTINGS_USERAPIENDPOINT: "https://sso.{{ .Values.domain }}/realms/uds/protocol/openid-connect/userinfo"
+        {{ end }}
+        {{ if eq .Values.sso.protocol "openid_connect" }}
+        MM_GITLABSETTINGS_ENABLE: "false"
```
Comment (Racer159) [inline chart/templates/uds-package.yaml:31]: why have this env var here?

File: chart/templates/uds-package.yaml [hunk 3]
```diff
@@ -14,16 +14,28 @@ spec:
         - "openid"
         - "mapper-oidc-username-username"
         - "mapper-oidc-mattermostid-id"
-        - "mapper-oidc-email-email" 
+        - "mapper-oidc-email-email"
 
       secretName: {{ .Values.sso.secretName }}
       secretTemplate:
+        {{ if eq .Values.sso.protocol "gitlab" }}
+        MM_OPENIDSETTINGS_ENABLE: "false"
```
Comment (Racer159) [inline chart/templates/uds-package.yaml:22]: why have this env var here?

**Review by Racer159 — CHANGES_REQUESTED** [review]

File: chart/templates/uds-package.yaml (hunk 2 above)
Comment (corang) [reply chart/templates/uds-package.yaml:31]: Was thinking in the case of moving from oidc -> gitlab or other way around I wanted to be explicit about disabling the other way of signing in

Comment (Racer159) [reply chart/templates/uds-package.yaml:31]: Wouldn't this secret get replaced in its entirety and then get merged into the pod spec (thus deleting the other env var set?)

Comment (corang) [reply chart/templates/uds-package.yaml:31]: Yeah, I'm worried about mattermost itself retaining settings, figured explicit disable was safer just in case

Comment (Racer159) [reply chart/templates/uds-package.yaml:31]: Confirmed that it flips without these env vars correctly - there is another issue with the config - will track it down
image

Comment (Racer159) [reply chart/templates/uds-package.yaml:31]: When changing sso providers though - MM tracks which users were created from which IDP so any users that used to login with `gitlab` will be disabled from logging in when switching to `openid_connect` based on matching their email address.

File: chart/templates/uds-package.yaml [hunk 4]
```diff
@@ -14,16 +14,28 @@ spec:
         - "openid"
         - "mapper-oidc-username-username"
         - "mapper-oidc-mattermostid-id"
-        - "mapper-oidc-email-email" 
+        - "mapper-oidc-email-email"
```
Comment (Racer159) [inline chart/templates/uds-package.yaml:17]: this needs to be an `if` statement to work correctly:

```
        {{ if eq .Values.sso.protocol "gitlab" }}
        - "mapper-oidc-username-username"
        - "mapper-oidc-mattermostid-id"
        - "mapper-oidc-email-email"
        {{ end }}
        {{ if eq .Values.sso.protocol "openid_connect" }}
        - "profile"
        - "email"
        {{ end }}
```

Comment (corang) [reply chart/templates/uds-package.yaml:17]: changed as requested

File: chart/templates/uds-package.yaml (hunk 2 above)
Comment (corang) [reply chart/templates/uds-package.yaml]: Removed unnecessary keys

File: chart/templates/uds-package.yaml [hunk 5]
```diff
@@ -11,19 +11,34 @@ spec:
       redirectUris:
         - "https://{{ .Values.subdomain }}.{{ .Values.domain }}/*"
       defaultClientScopes:
-        - "openid"
+        {{ if eq .Values.sso.protocol "gitlab" }}
```
Comment (Racer159) [inline chart/templates/uds-package.yaml:14]:
```suggestion
        - "openid"
        {{ if eq .Values.sso.protocol "gitlab" }}
```

Sorry I wasn't explicit - openid always needs to be there.

**Review by Racer159 — CHANGES_REQUESTED** [review]

**Review by Racer159 — APPROVED** [review]

**Pull request merged**